trusted/pal/src/pal.rs

```diff
@@ -0,0 +1,174 @@
+    /// For example, a trusted application that represents a node in Raft cluster
+    /// running inside of a group trusted hosts, it is important that Raft node
+    /// identity cannot be forged, does not require coordination to pick one and
+    /// has low collision chance. Hash of the public signing key is an identity
+    /// mechanism that is compliant with these requirements.
+    fn public_signing_key(&self) -> Vec<u8>;
+}
+
+/// Represents a trusted host, abstracting away the details of how trusted
+/// application is hosted (e.g. trusted host can be restricted kernel bare
+/// metal based encrypted virtual machine or a linux kernel docker based
+/// encrypted virtual machine).
+pub trait Host {
+    /// Gets attestation that represens a set of claims about the trusted application
+    /// and its host. The attestation is generated after the trusted application and
+    /// host are fully initialiezed, the attestation doesn't change throughout the
+    /// lifetime.
+    fn get_self_attestation(&self) -> Box<dyn Attestation>;
+
+    /// Gets serialized configuration that stays immutable through the lifetime of
+    /// the trusted application.
+    fn get_self_config(&self) -> Vec<u8>;
+
+    /// Sends messages through the communication channel that connects the trusted
+    /// application to the untrusted launcher.
+    ///
+    /// # Argumnets
+    ///
+    /// * `messages` - A set of messsages to send through the channel.
+    ///
+    /// # Returns
+    ///
+    /// Error if the communication channel is irrepairably broken, a success otherise.
+    fn send_messages(&mut self, messages: &[Message]) -> Result<()>;
+
+    /// Attempts to deserialize peer attestation and perform initial verification
+    /// of the attestation. The application specific verification (e.g. ensuring
+    /// that attestation points to a particular binary and configuration) is the
+    /// responsibility of the trusted application.
+    ///
+    /// # Arguments
+    ///
+    /// * `peer_attestation` - A serialized attestation of a peer trusted application.
+    ///
+    /// # Returns
+    ///
+    /// A deserialized and initially verified peer attestation if success, an error
+    /// otherwise. An error may be caused by malformed serialized representation or
+    /// by failing verification.
+    fn verify_peer_attestation(&self, peer_attestation: &[u8]) -> Result<Box<dyn Attestation>>;
+}
+
+/// Represents a trusted application running inside a trusted host. The trusted
+/// application is considered passive and performs execution in response to
+/// receiving messages through the communication channel that connects the trusted
+/// application to the untrusted launcher. In the absence of messages to be processed
+/// the trusted application will receive periodically empty set of messages to allow
+/// trusted application to make progress based on time change.
+pub trait Application {
```

Comment:
I think the application may also just implement the same [micro RPC transport trait](https://github.com/project-oak/oak/blob/8d2e369e8eda22eefd26eeef8994ff68b2541922/micro_rpc/src/lib.rs#L41-L49). It can just always return an empty message back.

Would that work?

Reply:
I think it is possible to implement the transport trait, however this will lead to two consequences:
- the access to the host (platform) will have to be exposed separately (say, through a separate init method on the application)
- the clock reading will have to be either exposed separately through the host (platform) or encapsulated into the messages themselves.
I feel like these two consequences will make the trait less clear and more complex.

Reply:
And of course, the microRPC is not necessarily the medium (in Docker as far as I understand the gRPC is the medium) which will break the abstraction.